File: microservices/realtime/src/Feeder/Feeder.php
<?php

namespace Feeder;

use Swoole\Coroutine;
use Swoole\Coroutine\Redis;

$loader = require __DIR__.'/../../vendor/autoload.php';

$helpers = new Class
{
    public function redisPush(Redis $redis, string $channel, array $payload = null)
    {
        $redis->publish(
            $channel,
            json_encode($payload)
        );

        if ($payload['Event'] === AbstractCall::CALL_SETUP) {
            echo $payload['Event'] . " " . $channel . "\n" . json_encode($payload);
        } else {
            echo $payload['Event'] . " " . $channel;
        }

        echo "\n";
    }

    public function progress(Redis $redis, AbstractCall $call)
    {
        $payload = $call->progress();
        $channel = key($payload);

        self::redisPush(
            $redis,
            $channel,
            $payload[$channel]
        );
    }
};

$trunksCalls = [];
$usersCalls = [];
$trunksCo = function () use ($trunksCalls, $usersCalls, $helpers) {

    $redis = new Redis();
    $redis->connect('127.0.0.1', 6379);

    $progessFn = function (&$calls, $newCallClass) use ($redis, $helpers) {

        if (empty($calls)) {

            /** @var AbstractCall $newCall */
            $newCall = new $newCallClass();
            $helpers->progress($redis, $newCall);
            $calls[$newCall->getChannel()] = $newCall;

            return;
        }

        $currentCallNum = count($calls);
        $increase = $currentCallNum < 5
            ? rand(1, 3) > 1
            : rand(1, 5) > 4;

        if ($increase) {
            /** @var AbstractCall $newCall */
            $newCall = new $newCallClass();
            $helpers->progress($redis, $newCall);
            $calls[$newCall->getChannel()] = $newCall;

            return;
        }

        $randomCallKey = array_rand($calls);

        /** @var AbstractCall $call */
        $call = $calls[$randomCallKey];
        $helpers->progress($redis, $call);

        if ($call->isDone()) {
            unset($calls[$randomCallKey]);
        }
    };

    while (true) {
        if (!empty($trunksCalls)) {
            echo
                "Current call number: "
                . (count($trunksCalls) + count($usersCalls))
                . "\n\n";
        }
        Coroutine::sleep(5);

        $progessFn(
            $trunksCalls,
            TrunksCall::class
        );

        echo "\n";

        $progessFn(
            $usersCalls,
            UsersCall::class
        );
    }
};

Coroutine::create($trunksCo);

File: microservices/realtime/pkg/services/feeder/feeder.go
package feeder

import (
	"fmt"
	"math/rand"
	"sync"
	"time"

	"github.com/redis/go-redis/v9"
	"irontec.com/realtime/pkg/config"
)

var trunksCalls map[string]*TrunksCall = make(map[string]*TrunksCall)
var usersCalls map[string]*UsersCall = make(map[string]*UsersCall)
var redisClient = redis.NewClient(&redis.Options{
	Addr: "127.0.0.1:6379",
	DB:   config.GetRedisDatabase(),
})
var wg sync.WaitGroup

func Execute() {

	trunksCo(trunksCalls, usersCalls)
}

func trunksCo(
	trunksCalls map[string]*TrunksCall,
	usersCalls map[string]*UsersCall,
) {

	for {
		fmt.Printf("Current call number: %d\n\n", len(trunksCalls)+len(usersCalls))

		time.Sleep(5 * time.Second)

		trunksCall := NewTrunksCall()
		progressTrunks(trunksCalls, trunksCall)

		fmt.Println()

		usersCall := NewUsersCall()
		progressUsers(usersCalls, usersCall)

	}

}

func progressTrunks(calls map[string]*TrunksCall, trunkCall *TrunksCall) {

	if len(calls) == 0 {

		newCall, err := ProgressTrunks(redisClient, trunkCall)

		if err != nil {
			fmt.Println("failed to progress call")
		}

		trunksCalls[newCall.GetChannel()] = newCall
		return
	}

	currentCallNum := len(calls)
	increase := currentCallNum < 5 && rand.Intn(3) > 1 || rand.Intn(5) > 4

	if increase {
		newCall, err := ProgressTrunks(redisClient, trunkCall)

		if err != nil {
			fmt.Println("failed to progress call")
		}

		trunksCalls[newCall.GetChannel()] = newCall
		return

	}

	randomCallKey := ""
	for key := range calls {
		randomCallKey = key
		break
	}

	call := calls[randomCallKey]
	call, err := ProgressTrunks(redisClient, call)

	if err != nil {
		fmt.Println("failed to progress call")
	}

	if call.IsDone() {
		delete(calls, randomCallKey)
	}

}

func progressUsers(calls map[string]*UsersCall, userCall *UsersCall) {

	if len(calls) == 0 {

		newCall, err := ProgressUsers(redisClient, userCall)

		if err != nil {
			fmt.Println("failed to progress call")
		}

		usersCalls[newCall.GetChannel()] = newCall
		return
	}

	currentCallNum := len(calls)
	increase := currentCallNum < 5 && rand.Intn(3) > 1 || rand.Intn(5) > 4

	if increase {

		newCall, err := ProgressUsers(redisClient, userCall)

		if err != nil {
			fmt.Println("failed to progress call")
		}

		usersCalls[newCall.GetChannel()] = newCall
		return
	}

	randomCallKey := ""
	for key := range calls {
		randomCallKey = key
		break
	}

	call := calls[randomCallKey]
	call, err := ProgressUsers(redisClient, call)

	if err != nil {
		fmt.Println("failed to progress call")
	}

	if call.IsDone() {
		delete(calls, randomCallKey)
	}

}
